### PR TITLE
Add `FileScanConfig::new()` API

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/10546

## Rationale for this change

While working on https://github.com/apache/datafusion/pull/10549 it was cumbersome to create a `FileScanConfig` -- there are many different fields, many of which need default values which require some thought to figure out.

The number of fields also makes it harder to see what is specifically set and what is the default

This has bothered me for a long time, so I made a PR to fix it.

You can see this in action here https://github.com/apache/datafusion/pull/10618

## What changes are included in this PR?

1. Add FileScanConfig::new() that takes required information
2. Add an example
2. Add builder style APIs for updating other fields
3. Update the code to use the new API

## Are these changes tested?
Covered by existing tests

## Are there any user-facing changes?
Nicer and easier to use API for creating FileScanConfig

### CODE REVIEW

File: datafusion-examples/examples/csv_opener.rs [hunk 1]
```diff
@@ -58,16 +57,11 @@ async fn main() -> Result<()> {
...
-        object_store_url: ObjectStoreUrl::local_filesystem(),
-        file_schema: schema.clone(),
-        file_groups: vec![vec![PartitionedFile::new(path.display().to_string(), 10)]],
-        statistics: Statistics::new_unknown(&schema),
-        projection: Some(vec![12, 0]),
-        limit: Some(5),
-        table_partition_cols: vec![],
-        output_ordering: vec![],
-    };
+    let scan_config =
```
Comment (alamb) [inline datafusion-examples/examples/csv_opener.rs:60]: This is a pretty good example of what the new API allows for:
1. Less code (defaults are set)
2. Makes it clearer what is the default and what is not

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs [hunk 2]
```diff
@@ -64,12 +64,43 @@ pub fn wrap_partition_value_in_dict(val: ScalarValue) -> ScalarValue {
...
+///
+/// // Read /tmp/file2.parquet 56 bytes and /tmp/file3.parquet 78 bytes
+/// // in a  single row group
+/// config.add_file_group(vec![
+///    PartitionedFile::new("file2.parquet", 56),
+///    PartitionedFile::new("file3.parquet", 78),
+/// ]);
+/// ```
 #[derive(Clone)]
 pub struct FileScanConfig {
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:93]: Here is the new API and documentation. Note this is entirely backwards compatible (I did not change any fields to non pub, etc)

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs [hunk 3]
```diff
@@ -101,6 +133,99 @@ pub struct FileScanConfig {
...
+        self
+    }
+
+    /// Set the limit of the files
+    pub fn with_limit(mut self, limit: Option<usize>) -> Self {
+        self.limit = limit;
+        self
+    }
+
+    /// Add a file as a single group
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:190]: Having to wrap a single file in a `vec![vec![..]]` makes sense from the implementation point of view, but most users shouldn't have to worry about this

**Review by phillipleblanc — APPROVED** [review]
Very clean, thank you! This is the first time I've seen both the `with_<x>` and `add_<x>` pattern used on a builder-like struct.

Is that something you generally do for all builder-type APIs, or it depends?

**Comment by alamb** [comment]
Thank you @phillipleblanc 

> Is that something you generally do for all builder-type APIs, or it depends?

I would say it depends. In this case I hadd a use case for `add` style in the example I am working on for indexing (you can see this in action here https://github.com/apache/datafusion/pull/10618)

However, I can see how we could do something else in this case, so perhaps I should remove the `set` style APIs 🤔

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs [hunk 4]
```diff
@@ -64,12 +64,43 @@ pub fn wrap_partition_value_in_dict(val: ScalarValue) -> ScalarValue {
...
 /// The base configurations to provide when creating a physical plan for
 /// any given file format.
+///
+/// # Example
+/// ```
+/// # use std::sync::Arc;
+/// # use arrow_schema::Schema;
+/// use datafusion::datasource::listing::PartitionedFile;
+/// # use datafusion::datasource::physical_plan::FileScanConfig;
+/// # use datafusion_execution::object_store::ObjectStoreUrl;
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:74]: 🤔  maybe I should change this example to use the with style 🤔

**Review by alamb — COMMENTED** [review]
@phillipleblanc  upon more consideration I htink you are right that sticking to one pattern (e.g. builder) will be cleaner. I removed the `add*` apis in f228af32d

**Review by metegenez — APPROVED** [review]
Thanks for the effort, this was a necessary API to introduce.

**Comment by alamb** [comment]
Thank you @metegenez

**Comment by alamb** [comment]
Thanks everyone for the reviews!

**Pull request merged**